packages/app/providers/wallet-mobile-sdk-provider.tsx

```diff
@@ -0,0 +1,120 @@
+import { useCallback, useRef, useState, useEffect } from "react";
+
+import {
+  initiateHandshake,
+  makeRequest,
+  isConnected,
+  resetSession,
+} from "@coinbase/wallet-mobile-sdk";
+import { ethers } from "ethers";
+import { MMKV } from "react-native-mmkv";
+
+import { WalletMobileSDKContext } from "../context/wallet-mobile-sdk-context";
+
+const CB_WALLET_METADATA = {
+  name: "Coinbase Wallet",
+};
+
+const CACHED_ADDRESS_KEY = "mobile_sdk.address";
+const storage = new MMKV();
+
+type OnConnectedListener = (success: boolean) => void;
+
+interface WalletMobileSDKProviderProps {
+  children: React.ReactNode;
+}
+
+export function WalletMobileSDKProvider({
+  children,
+}: WalletMobileSDKProviderProps) {
+  const onConnectedListenerRef = useRef<OnConnectedListener | null>(null);
+  const [connected, setConnected] = useState<boolean>(false);
+  const [address, setAddress] = useState<string | null>(null);
+
+  const connect = useCallback(async () => {
+    try {
+      const action = {
+        method: "eth_requestAccounts",
+        params: {},
+      };
+      const [, account] = await initiateHandshake([action]);
+
+      const callback = onConnectedListenerRef.current;
+      if (account && account.address) {
+        setConnected(true);
+        setAddress(account.address);
+        storage.set(CACHED_ADDRESS_KEY, account.address);
+
+        callback && callback(true);
+        onConnectedListenerRef.current = null;
+      } else {
+        callback && callback(false);
+        onConnectedListenerRef.current = null;
+      }
+    } catch (error) {
+      resetSession();
+      setAddress(null);
+      setConnected(false);
+      onConnectedListenerRef.current = null;
+    }
+  }, []);
+
+  const disconnect = useCallback(() => {
+    resetSession();
+    setConnected(false);
+    setAddress(null);
+    storage.delete(CACHED_ADDRESS_KEY);
+  }, []);
+
+  const onConnected = useCallback(() => {
+    return new Promise<void>((resolve, reject) => {
+      onConnectedListenerRef.current = (success) => {
+        if (success) {
+          resolve();
```

Comment:
nit: not very sure I think we could move `onConnectedListenerRef.current = null;` after resolve/reject calls so we don't have to repeat?